### ISSUE TITLE
[iOS] Does not run next to other SDKs using BGTaskScheduler

### ISSUE THREAD

**Issue by flodaniel** [transistorsoft/flutter_background_geolocation#1205]
## Your Environment
* Plugin version: 4.11
* Platform: iOS
* OS version: iOS 16
* Device manufacturer / model:
* Flutter info (`flutter doctor`): 
* Plugin config:
```dart <-- Syntax highlighting: DO NOT REMOVE -->
    <key>BGTaskSchedulerPermittedIdentifiers</key>
    <array>
        <string>com.transistorsoft.fetch</string>
        <string>com.sentiance.backgroundtask.tilerefresh</string>
        <string>com.sentiance.backgroundtask.task_processing</string>
    </array>
```

## Expected Behavior

I would expect the app to still run when I add `com.transistorsoft.fetch` to my `BGTaskSchedulerPermittedIdentifiers` list.

## Actual Behavior

Instead i see a crash in my xcode log:

<details>
	<summary>Crash Log</summary>

``` 
```
2023-06-28 11:45:29.751489+0200 Runner[49390:28644861] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Launch handler for task with identifier com.sentiance.backgroundtask.task_processing has already been registered'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000115afe78b __exceptionPreprocess + 242
	1   libobjc.A.dylib                     0x000000011157cb73 objc_exception_throw + 48
	2   Foundation                          0x0000000111f51f64 _userInfoForFileAndLine + 0
	3   BackgroundTasks                     0x00000001108f3b52 -[BGTaskScheduler _unsafe_registerForTaskWithIdentifier:usingQueue:launchHandler:] + 366
	4   BackgroundTasks                     0x00000001108f39aa -[BGTaskScheduler registerForTaskWithIdentifier:usingQueue:launchHandler:] + 175
	5   Runner                              0x0000000100ae4863 __47-[SENTBackgroundTaskScheduler registerAllTasks]_block_invoke + 733
	6   Runner                              0x0000000100ae44c9 -[SENTBackgroundTaskScheduler registerAllTasks] + 436
	7   Runner                              0x0000000100bfb79a __43-[Sentiance(Private) _setupBackgroundTasks]_block_invoke + 55
	8   libdispatch.dylib                   0x0000000113c9c7ec _dispatch_client_callout + 8
	9   libdispatch.dylib                   0x0000000113c9de85 _dispatch_once_callout + 66
	10  Runner                              0x0000000100be9710 __79-[Sentiance _initializeWithOptions:launchOptions:isReinitializationAfterReset:]_block_invoke + 866
	11  Runner                              0x0000000100b86ed7 __63-[SENTSdkInternalInitState lockStateForWritingAndDispatchSync:]_block_invoke + 61
	12  libdispatch.dylib                   0x0000000113c9c7ec _dispatch_client_callout + 8
	13  libdispatch.dylib                   0x0000000113cae70b _dispatch_lane_barrier_sync_invoke_and_complete + 132
	14  Runner                              0x0000000100b86df1 -[SENTSdkInternalInitState lockStateForWritingAndDispatchSync:] + 192
	15  Runner                              0x0000000100be9243 -[Sentiance _initializeWithOptions:launchOptions:isReinitializationAfterReset:] + 326
	16  Runner                              0x00000001010adbc8 $s13sentiance_ios20SwiftSentiancePluginC11application_29didFinishLaunchingWithOptionsSbSo13UIApplicationC_SDys11AnyHashableVypGtF + 376
	17  Runner                              0x00000001010adf7f $s13sentiance_ios20SwiftSentiancePluginC11application_29didFinishLaunchingWithOptionsSbSo13UIApplicationC_SDys11AnyHashableVypGtFTo + 95
	18  Flutter                             0x000000011fd54310 -[FlutterPluginAppLifeCycleDelegate application:didFinishLaunchingWithOptions:] + 282
	19  Runner                              0x00000001003cc86e $s6Runner11AppDelegateC11application_29didFinishLaunchingWithOptionsSbSo13UIApplicationC_SDySo0j6LaunchI3KeyaypGSgtF + 2142
	20  Runner                              0x00000001003ccb87 $s6Runner11AppDelegateC11application_29didFinishLaunchingWithOptionsSbSo13UIApplicationC_SDySo0j6LaunchI3KeyaypGSgtFTo + 183
	21  UIKitCore                           0x0000000132ed62c3 -[UIApplication _handleDelegateCallbacksWithOptions:isSuspended:restoreState:] + 271
	22  UIKitCore                           0x0000000132ed8186 -[UIApplication _callInitializationDelegatesWithActions:forCanvas:payload:fromOriginatingProcess:] + 4288
	23  UIKitCore                           0x0000000132edded2 -[UIApplication _runWithMainScene:transitionContext:completion:] + 1236
	24  UIKitCore                           0x00000001322f723e -[_UISceneLifecycleMultiplexer completeApplicationLaunchWithFBSScene:transitionContext:] + 122
	25  UIKitCore                           0x0000000132975510 _UIScenePerformActionsWithLifecycleActionMask + 88
	26  UIKitCore                           0x00000001322f7dd9 __101-[_UISceneLifecycleMultiplexer _evalTransitionToSettings:fromSettings:forceExit:withTransitionStore:]_block_invoke + 198
	27  UIKitCore                           0x00000001322f781a -[_UISceneLifecycleMultiplexer _performBlock:withApplicationOfDeactivationReasons:fromReasons:] + 252
	28  UIKitCore                           0x00000001322f7c22 -[_UISceneLifecycleMultiplexer _evalTransitionToSettings:fromSettings:forceExit:withTransitionStore:] + 846
	29  UIKitCore                           0x00000001322f741b -[_UISceneLifecycleMultiplexer uiScene:transitionedFromState:withTransitionContext:] + 354
	30  UIKitCore                           0x0000000132306529 __186-[_UIWindowSceneFBSSceneTransitionContextDrivenLifecycleSettingsDiffAction _performActionsForUIScene:withUpdatedFBSScene:settingsDiff:fromSettings:transitionContext:lifecycleActionType:]_block_invoke + 178
	31  UIKitCore                           0x0000000132842703 +[BSAnimationSettings(UIKit) tryAnimatingWithSettings:fromCurrentState:actions:completion:] + 833
	32  UIKitCore                           0x00000001329935b0 _UISceneSettingsDiffActionPerformChangesWithTransitionContextAndCompletion + 257
	33  UIKitCore                           0x00000001323061a9 -[_UIWindowSceneFBSSceneTransitionContextDrivenLifecycleSettingsDiffAction _performActionsForUIScene:withUpdatedFBSScene:settingsDiff:fromSettings:transitionContext:lifecycleActionType:] + 348
	34  UIKitCore                           0x00000001320e8eed __64-[UIScene scene:didUpdateWithDiff:transitionContext:completion:]_block_invoke.601 + 815
	35  UIKitCore                           0x00000001320e7903 -[UIScene _emitSceneSettingsUpdateResponseForCompletion:afterSceneUpdateWork:] + 245
	36  UIKitCore                           0x00000001320e8a9a -[UIScene scene:didUpdateWithDiff:transitionContext:completion:] + 255
	37  UIKitCore                           0x0000000132edc6d8 -[UIApplication workspace:didCreateScene:withTransitionContext:completion:] + 733
	38  UIKitCore                           0x000000013287511f -[UIApplicationSceneClientAgent scene:didInitializeWithEvent:completion:] + 350
	39  FrontBoardServices                  0x0000000118c8fbea -[FBSScene _callOutQueue_agent_didCreateWithTransitionContext:completion:] + 415
	40  FrontBoardServices                  0x0000000118cbec0e __92-[FBSWorkspaceScenesClient createSceneWithIdentity:parameters:transitionContext:completion:]_block_invoke.187 + 102
	41  FrontBoardServices                  0x0000000118c9dbfc -[FBSWorkspace _calloutQueue_executeCalloutFromSource:withBlock:] + 209
	42  FrontBoardServices                  0x0000000118cbe815 __92-[FBSWorkspaceScenesClient createSceneWithIdentity:parameters:transitionContext:completion:]_block_invoke + 344
	43  libdispatch.dylib                   0x0000000113c9c7ec _dispatch_client_callout + 8
	44  libdispatch.dylib                   0x0000000113ca01c0 _dispatch_block_invoke_direct + 498
	45  FrontBoardServices                  0x0000000118ce5507 __FBSSERIALQUEUE_IS_CALLING_OUT_TO_A_BLOCK__ + 30
	46  FrontBoardServices                  0x0000000118ce53fd -[FBSSerialQueue _targetQueue_performNextIfPossible] + 174
	47  FrontBoardServices                  0x0000000118ce552f -[FBSSerialQueue _performNextFromRunLoopSource] + 19
	48  CoreFoundation                      0x0000000115a5cb8f __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
	49  CoreFoundation                      0x0000000115a5cad1 __CFRunLoopDoSource0 + 157
	50  CoreFoundation                      0x0000000115a5c2cd __CFRunLoopDoSources0 + 217
	51  CoreFoundation                      0x0000000115a569ba __CFRunLoopRun + 889
	52  CoreFoundation                      0x0000000115a56264 CFRunLoopRunSpecific + 560
	53  GraphicsServices                    0x000000011c5d924e GSEventRunModal + 139
	54  UIKitCore                           0x0000000132eda7bf -[UIApplication _run] + 994
	55  UIKitCore                           0x0000000132edf5de UIApplicationMain + 123
	56  Runner                              0x00000001003cd55f main + 63
	57  dyld                                0x000000010a701384 start_sim + 10
	58  ???                                 0x00000002021af310 0x0 + 8625255184
)
libc++abi: terminating due to uncaught exception of type NSException

```
```

</details>

## Steps to Reproduce

1. Add plugin to pubspec.yml
2. Add `com.transistorsoft.fetch` to `BGTaskSchedulerPermittedIdentifiers`
3. Add other SDK (like Sentiance SDK, which also does location tracking stuff) and its required `BGTaskSchedulerPermittedIdentifiers`: `com.sentiance.backgroundtask.tilerefresh` and `com.sentiance.backgroundtask.task_processing`
4. Compile and try to launch app
5. Observe crash

## Context

Trying to get the plugin to run next to the sentiance SDK as we are evaluating how `flutter_background_location` plugin could be an addition to the tracking capabilities of our app. 

## Debug logs

<details>
	<summary>Logs</summary>

``` <!-- syntax-highligting:  DO NOT REMOVE -->
see above
```

</details>

**Comment by christocracy** [comment]
The iOS BGTaskScheduler api is not easy to share between multiple components.

Background Fetch automatically consumes all strings found in your plist and attempts to register them to itself.

You will have to completely disable using background_fetch in favour of this "sentience" sdk.

**Comment by flodaniel** [comment]
Ok thanks for the info.
Will your geofencing capabilitites still work, especially in a terminated app state (killed through app switcher)?

**Comment by christocracy** [comment]
Background Geolocation doesn't require background_fetch to operate.  It's just a "nice to have"

**Comment by flodaniel** [comment]
Awesome, thanks a lot!
We will then restart our testing with what is possible without purchasing a license. Once we have it working we are looking forward to getting the license!

**Comment by flodaniel** [comment]
One more follow-up question. The docs do not state for what background_fetch is used: https://github.com/transistorsoft/flutter_background_geolocation/blob/master/help/INSTALL-IOS.md#configure-background_fetch. 

Is there additional docs to understand the limitations of not using it? 
Also the docs there state it is absolutely required.

**Comment by christocracy** [comment]
It is used for the scheduler, which you're probably not using.

It is used to periodically check its database for records to upload, which may exist there due to lack of network connection previously.

That's about it.

**Comment by flodaniel** [comment]
Ok makes sense. 
I don't think so! If it works we will also use more of the location tracking features most likely in foreground/background but not headless and always requiring some user interaction :)

**Comment by flodaniel** [comment]
I have to reopen this issue as the crash still occurs even without adding `com.transistorsoft.fetch` to `BGTaskSchedulerPermittedIdentifiers`. I suspect your `background_fetch` calls `registerTaskWithIdentifier` by default, even when not used? 


```
*** Assertion failure in -[BGTaskScheduler _unsafe_registerForTaskWithIdentifier:usingQueue:launchHandler:], BGTaskScheduler.m:189

*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Launch handler for task with identifier com.sentiance.backgroundtask.task_processing has already been registered'
```

**Comment by flodaniel** [comment]
With my limited iOS knowledge i suspect that the problem comes from didFinishLaunching here: https://github.com/transistorsoft/transistor-background-fetch/blob/6feb11c93f4e5debf7792e7e7c444f3d62ac733c/ios/TSBackgroundFetch/TSBackgroundFetch/TSBackgroundFetch.m#L54C1-L65C2.

Is it possible to limit the loop to only run for identifiers starting with `com.transistorsoft.` ?
Based on the [current docs](https://github.com/transistorsoft/flutter_background_fetch/blob/master/help/INSTALL-IOS.md) this should not be breaking change, as custom tasks also need to start with `com.transistorsoft.`

**Comment by christocracy** [comment]
I have [added code to `TSBackgroundFetch`](https://github.com/transistorsoft/transistor-background-fetch/blob/0dd1457588ba031f11e9452f125144aac8a9b75f/ios/TSBackgroundFetch/TSBackgroundFetch/TSBackgroundFetch.m#L62-L72) to enforce a prefix of `com.transistorsoft` for custom tasks.

Test it out by installing from the branch `prefix-processing-task` as follows:

```yaml
dependencies:
  background_fetch:
    git:
      url: https://github.com/transistorsoft/flutter_background_fetch
      ref: prefix-processing-task
```

**Comment by flodaniel** [comment]
This worked. I added it as a dependency override.
Also called `ready` and `addGeofence` without any other crash.

**Comment by flodaniel** [comment]
I see this is released in 1.2.2 https://pub.dev/packages/background_fetch/changelog
